Add Alfawise Humidifier

HydrelioxGitHub opened the pull request:
## Description:
Add Alfawise Humidifier as fan platform

This is the clean PR coming from the #13923 wip.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5192

## Example entry for `configuration.yaml` (if applicable):
```yaml
fan:
 - platform: alfawise
   host: 192.168.1.101
   name: Living Room
   mac: AABBCCDDEEFFF
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

HydrelioxGitHub commented:
It seems that Travis is showing errors that's not connected to this PR... What should I do ?

fabaff commented on homeassistant/components/fan/alfawise.py:
```diff
@@ -0,0 +1,121 @@
+from homeassistant.components.fan import (SPEED_LOW, SPEED_HIGH,
+                                          FanEntity, SUPPORT_SET_SPEED)
+from homeassistant.const import CONF_DEVICES, CONF_NAME, CONF_MAC
+from homeassistant.const import STATE_OFF
```
`STATE_OFF` can be imported with the previous import.

fabaff commented on homeassistant/components/fan/alfawise.py:
```diff
@@ -0,0 +1,121 @@
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {vol.Required(CONF_DEVICES): {isip: DEVICE_SCHEMA}})
+
```
The configuration should be something like:

```yaml
  - platform: alfawise
    host: 192.168.1.101
    name: Living Room
    mac: AABBCCDDEEFF
```
Easier to move around in the configuration and less error-prone. I guess that most users will not have more than two devices unlike `template` platforms.

HydrelioxGitHub replied:
I've changed the configuration to match your suggestion. If the PR is merged I  will update the documentation as well.

fabaff commented on homeassistant/components/fan/alfawise.py:
```diff
@@ -0,0 +1,121 @@
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {vol.Required(CONF_DEVICES): {isip: DEVICE_SCHEMA}})
```
It shouldn't be limited to IP addresses. The devices could have a DNS name depending on the configuration.

HydrelioxGitHub replied:
This is a limit of the third party library. IP must be provided.

fabaff commented on homeassistant/components/fan/alfawise.py:
```diff
@@ -0,0 +1,121 @@
+REQUIREMENTS = ['pyAlfawise==0.5-beta']
+
+
+def isip(value: Any) -> str:
```
Checks like this should be part of `config_validation.py` because it could be re-used by other platforms.

HydrelioxGitHub replied:
This is what I asked to @MartinHjelmare in the previous PR #13923  . He tells me to put it in the platform file instead of the config_validation.py file

fabaff replied:
The problem is that if it part of the platform then the validator is not documented and the chance is high that somebody is doing the same again. Also, it's not tested.

MartinHjelmare replied:
We usually wait with adding a new validator to the helper module until at least two other modules require the validator.

But it's not a big deal.

balloob commented on homeassistant/helpers/config_validation.py:
```diff
@@ -446,6 +447,19 @@ def socket_timeout(value):
             raise vol.Invalid('Invalid socket timeout: {err}'.format(err=_))
 
 
+def isip(value: Any) -> str:
```
Please remove this and instead use `from ipaddress import ip_address` and use `ip_address` as a validator.

HydrelioxGitHub committed:
Add Alfawise Humidifier 

Added as fan platform
Correction

Move ip validator to helper
Correction
Change ip validator with correct import and return type

HydrelioxGitHub commented:
Am I missing something to correct on this P.R. ? Or is it just a lack of time to merge ?

MartinHjelmare commented on homeassistant/components/fan/alfawise.py:
```diff
@@ -0,0 +1,102 @@
+    """Set up the Alfawise fan platform."""
+    import pyAlfawise
+    fans = []
+    for ipaddr, device_config in config[CONF_DEVICES].items():
```
This doesn't match the config schema.

MartinHjelmare reviewed (changes requested):
Are the latest changes tested?

MartinHjelmare commented:
Please don't squash commits after review has started.
